src/vm_cpu.rs

```diff
@@ -96,6 +100,246 @@ pub struct VmCpuTrapState {
+            match state {
+                // The Started(_) handles the corner case where some counters are automatically
+                // started on configuration, and the subsequent call to start_counters
+                // fails with "already started" (example: insret).
+                Started(c) | Configured(c) => {
+                    *state = Started(*c);
+                }
+                _ => unreachable!(),
+            }
+        }
+    }
+
+    /// Updates internal state for PMU counters.
+    /// This should be called following a successful SBI call to stop counters.
+    pub fn update_stopped_counters(
+        &mut self,
+        counter_index: LogicalCounterIndex,
+        counter_mask: LogicalCounterMask,
+        stop_flags: PmuCounterStopFlags,
+    ) {
+        use PmuCounterState::*;
+        // Unwrap ok: counter_index and counter_mask were previously validated
+        let bitmask_iter = CounterMaskIter::new(counter_index.0, counter_mask.0).unwrap();
+        for i in bitmask_iter {
+            let state = &mut self.counter_state[i];
+            match state {
+                // Deliberately more permissive since the implementation permits
+                // operations even on stopped counters (example: stop_flag_reset).
+                Configured(c) | Started(c) => {
+                    if stop_flags.get_reset_flag() {
+                        *state = NotConfigured;
+                    } else {
+                        *state = PmuCounterState::Configured(*c);
+                    }
+                }
+                _ => unreachable!(),
+            }
+        }
+    }
+
+    /// Returns if the PMU counter range can be started.
+    pub fn is_counter_range_startable(
+        &self,
+        counter_index: LogicalCounterIndex,
+        counter_mask: LogicalCounterMask,
+    ) -> SbiResult<()> {
+        use PmuCounterState::*;
+        let mut bitmask_iter = CounterMaskIter::new(counter_index.0, counter_mask.0)?;
+        let result = bitmask_iter.any(|i| matches!(self.counter_state[i], Configured(_)));
+        match result {
+            true => Ok(()),
+            false => Err(SbiError::InvalidParam),
+        }
+    }
+
+    /// Returns if the PMU counter range can be stopped.
+    pub fn is_counter_range_stoppabble(
+        &self,
+        counter_index: LogicalCounterIndex,
+        counter_mask: LogicalCounterMask,
+    ) -> SbiResult<()> {
+        use PmuCounterState::*;
+        let mut bitmask_iter = CounterMaskIter::new(counter_index.0, counter_mask.0)?;
+        let result = bitmask_iter.any(|i| {
+            // Deliberately more permissive since the implementation permits
+            // operations even on stopped counters (example: stop_flag_reset).
+            matches!(self.counter_state[i], Started(_) | Configured(_))
+        });
+        match result {
+            true => Ok(()),
+            false => Err(SbiError::InvalidParam),
+        }
```

Comment:
Instead of using `any()` and matching against a bool, how about:

```
bitmask_iter.find(|i| ...).ok_or(SbiError::InvalidParam)?;
Ok(())
```

Same for the other `is_counter_range_*` methods.

Reply:
Think your comment holds if any is used, but it was a typo, and has been replaced with all.